pkg/controller/networkpolicy/status_controller.go

```diff
@@ -0,0 +1,343 @@
+	for {
+		select {
+		case event, ok := <-resultCh:
+			if !ok {
+				return
+			}
+			// Skip handling Bookmark events.
+			if event.Type == watch.Bookmark {
+				continue
+			}
+			np := event.Object.(*controlplane.NetworkPolicy)
+			if np.SourceRef.Type == controlplane.K8sNetworkPolicy {
+				continue
+			}
+			c.queue.Add(np.Name)
+		}
+	}
+}
+
+func (c *StatusController) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (c *StatusController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncHandler(key.(string))
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	klog.Errorf("Failed to sync NetworkPolicy status %s: %v", key, err)
+	c.queue.AddRateLimited(key)
+
+	return true
+}
+
+// syncHandler calculates the NetworkPolicy status based on the desired state from the internalNetworkPolicyStore and
+// the actual state from the statuses map, and syncs it with the Kubernetes API.
+func (c *StatusController) syncHandler(key string) error {
+	klog.V(2).Infof("Syncing NetworkPolicy status for %s", key)
+	internalNPObj, found, _ := c.internalNetworkPolicyStore.Get(key)
+	if !found {
+		// It has been deleted, cleaning its statuses.
+		c.clearStatuses(key)
+		return nil
+	}
+	internalNP := internalNPObj.(*antreatypes.NetworkPolicy)
+
+	// It means the NetworkPolicy hasn't been processed once. Set it to Pending to differentiate from NetworkPolicies
+	// with a span of 0 Node.
+	if internalNP.SpanMeta.NodeNames == nil {
+		return c.npControlInterface.UpdateAntreaNetworkPolicyStatus(internalNP.SourceRef.Namespace, internalNP.SourceRef.Name, &secv1alpha1.NetworkPolicyStatus{
+			Phase:              secv1alpha1.NetworkPolicyPending,
+			ObservedGeneration: internalNP.Generation,
+		})
+	}
+	desiredNodes := len(internalNP.SpanMeta.NodeNames)
+	currentNodes := 0
+	statuses := c.getNodeStatuses(key)
+	for _, status := range statuses {
+		// The node is no longer in the span of this policy, delete its status.
+		if !internalNP.NodeNames.Has(status.NodeName) {
+			c.deleteNodeStatus(key, status.NodeName)
+			continue
+		}
+		if status.Generation == internalNP.Generation {
+			currentNodes += 1
+		}
+	}
+
+	phase := secv1alpha1.NetworkPolicyRealizing
+	if currentNodes == desiredNodes {
+		phase = secv1alpha1.NetworkPolicyRealized
+	}
+
+	status := &secv1alpha1.NetworkPolicyStatus{
+		Phase:                phase,
+		ObservedGeneration:   internalNP.Generation,
+		CurrentNodesRealized: int32(currentNodes),
+		DesiredNodesRealized: int32(desiredNodes),
+	}
+	klog.V(2).Infof("Updating NetworkPolicy %s status: %v", internalNP.SourceRef.ToString(), status)
+	if internalNP.SourceRef.Type == controlplane.AntreaNetworkPolicy {
+		return c.npControlInterface.UpdateAntreaNetworkPolicyStatus(internalNP.SourceRef.Namespace, internalNP.SourceRef.Name, status)
```

Comment:
Will it be a problem to make too many K8s calls, when the span is big?

Reply:
The times of K8s calls should be consistent regardless the span of the policy because each round of syncHandler processes all statuses it gets so far. Say policy A applies to 10 Nodes, policy B applies to 100 Nodes.
Policy A got 3 Node updates first, it sets the policy status to 3/10 in the first round. During its first processing, it receives 5 updates, it then sets the policy status to 8/10 in the second round. During its second processing, it receives remaining updates, it sets status to 10/10 in the third round.
Policy B got 2 Node updates first, it sets the policy status to 2/100 in the first round. During its first processing, it receives 30 updates, it then sets the policy status to 32/100 in the second round. During its second processing, it receives 40 updates, it sets status to 72/100 in the third round, e.g..
This is similar to how Deployment/DaemonSet/Endpoints's status is updated.

Reply:
Sorry I might have missed something. If you have 10 Nodes, and they update status one by one, happen to after the controller processes the last update, then we finally we should update CRD status 10 times?

Reply:
If every update happens to be received after the controller processes the last update, it will update CRD status 10 times. But this shouldn't happen in practice. Every node should receive and process the policy at almost the same time, so the realization statuses should be reported closely. And updating CRD status could take some time, during which status reports from agents can be accumulated. There is also a limited number of workers, if there are many policies need to be sync, each policy won't get processed immediately once their statuses are received, which also make some time window to accumulate the status reports.

Reply:
Do you think you can describe your theory in the some comments in code? I am not very certain on it, and like to note it down for later attention.

Reply:
added to comments of syncHandler.